### PR TITLE
Add tests for Paillier encryption and decryption

### LINKED ISSUE #100: Add basic tests for the Paillier module
The Paillier module doesn't have any tests about behavior of `encrypt()` and `decrypt()`. Add all the reasonable unit tests you can think of. Sample list:
1. encrypting, then immediately decrypting, works.
2. decrypting with the wrong key fails
3. changing one bit of the encrypted blob causes decryption to fail
4. decrypting a random blob fails? I'm actually not sure if this is a required property.
5. input to encryption has to be in `Z_N`

### LINKED ISSUE #101: Clean up Paillier encryption decryption API
The `decrypt` function currently returns a `Vec<u8>`, but in every place we decrypt something, we call `BigNumber::from_slice` on the result. Pull that function call into the Paillier API so `decrypt` just returns a `BigNumber`.

Maybe do this before or with #100 so it's easier to write those tests.

### PR DESCRIPTION
**Pull request by marsella**
Closes #100. Closes #101.

This adds some fairly standard tests for Paillier encryption + decryption.

It also addresses something I'm not sure is a practical issue: the paper says
> We represent integers modulo N in the interval {−N/2, . . . , N/2} (rather than the canonical representation); this convention is crucial to the security analysis.

I'm not sure if this is also crucial to the _practical_ security of the library, but we encrypt a lot of things drawn from a range like `+/- 2^ell`. I wanted some kind of bounds check on encrypted messages, so I check that they're in the `[-N/2, N/2]` range. Then, to make decryption correct in the most reasonable way, I convert the output from the default canonical representation `[0, N)` to the range used in the protocol.

### CODE REVIEW

File: src/paillier.rs [hunk 1]
```diff
@@ -86,11 +105,21 @@ impl PaillierEncryptionKey {
...
             .0
             .decrypt(&c.0)
-            .ok_or(PaillierError::DecryptionFailed)?)
+            .ok_or(PaillierError::DecryptionFailed)
+            .map(BigNumber::from_slice)?;
+
+        // Switch representation into `[-N/2, N/2]`. libpaillier (and indeed, `BigNumber`s
+        // in general) returns values represented in the canonical range [0, N). A single
+        // subtraction will land us in the expected range for this application.
+        if x > (self.0.n() - 1) / 2 {
```
Comment (amaloz) [inline src/paillier.rs:118]: If the above `half_n` suggestion is taken, we can replace this with `x > self.half_n`

File: src/paillier.rs [hunk 2]
```diff
@@ -272,4 +306,133 @@ mod test {
             paillier_keygen_produces_good_primes()
         }
     }
+
+    fn random_message(
+        rng: &mut (impl CryptoRng + RngCore),
+        encryption_key: &PaillierEncryptionKey,
+    ) -> BigNumber {
+        let range: BigNumber = encryption_key.n() / 2;
```
Comment (amaloz) [inline src/paillier.rs:314]: Should this be `(n - 1) / 2`? Depending on how `BigNumber` rounds it might affect things.

File: src/paillier.rs [hunk 3]
```diff
@@ -272,4 +306,133 @@ mod test {
...
+        // Test a number in between N/2 and N (both + and -)
+        let too_big = (encryption_key.n() / 3) * 2;
+        assert!(encryption_key.encrypt(&mut rng, &too_big).is_err());
+        assert!(encryption_key.encrypt(&mut rng, &-too_big).is_err());
+
+        // Test a number bigger than N (both + and -)
+        let way_too_big = encryption_key.n() + 1;
+        assert!(encryption_key.encrypt(&mut rng, &way_too_big).is_err());
+        assert!(encryption_key.encrypt(&mut rng, &-way_too_big).is_err());
+    }
```
Comment (amaloz) [inline src/paillier.rs:378]: Might be good to add tests right at the edge of the boundary to make sure there's no off-by-one happening.

**Review by amaloz — COMMENTED** [review]
Looks good! A couple of minor comments.

File: src/paillier.rs [hunk 4]
```diff
@@ -66,11 +70,26 @@ impl PaillierEncryptionKey {
...
+    /// `[-N/2, N/2]`. Encryption will fail if `x` is outside this range.
     pub(crate) fn encrypt<R: RngCore + CryptoRng>(
         &self,
         rng: &mut R,
         x: &BigNumber,
     ) -> Result<(PaillierCiphertext, PaillierNonce)> {
+        // Compute the floor of n/2. Since `n` is the product of two primes, it'll be odd, so we
+        // can do this to make sure we get the actual floor. BigNumber division doesn't document
+        // whether it truncates or rounds for integer division.
+        let half_n: BigNumber = (self.n() - 1) / 2;
```
[the remark this answers is not in the document]
Comment (marsella) [reply src/paillier.rs:86]: I added a function to compute it, but I feel hesitant about optimizing for performance here before we have deserialization validation. I don't want `half_n` to get out of sync with actual `n`.

File: src/paillier.rs (hunk 1 above)
Comment (marsella) [reply src/paillier.rs:118]: Sadly this is on `DecryptionKey`, not `EncryptionKey`, so I made another function to do the same thing for this type.

File: src/paillier.rs (hunk 2 above)
Comment (marsella) [reply src/paillier.rs:314]: Good catch, yes!

File: src/paillier.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (amaloz) [reply src/paillier.rs:86]: Makes sense. Should we mark this for something to revisit when we get to deserialization validation or just leave it as is?

File: src/paillier.rs (hunk 3 above)
Comment (marsella) [reply src/paillier.rs:410]: Added!

File: src/paillier.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (marsella) [reply src/paillier.rs]: Added a note to #105

**Comment by marsella** [comment]
@amaloz This is ready for re-review.

File: src/paillier.rs [hunk 5]
```diff
@@ -59,25 +63,52 @@ impl PaillierCiphertext {
...
+}
 
 impl PaillierEncryptionKey {
     pub(crate) fn n(&self) -> &BigNumber {
-        self.0.n()
+        self.pk.n()
+    }
+
+    /// Get a reference to the internal `libpaillier` encryption key.
+    pub(crate) fn as_encryption_key(&self) -> &libpaillier::EncryptionKey {
```
Comment (amaloz) [inline src/paillier.rs:76]: #64 will get rid of the need to manipulate the `libpaillier::EncryptionKey` directly in `participant.rs`, so I don't think this code is needed (or we can just remove it when that PR gets merged in).

**Review by amaloz — APPROVED** [review]
Looks good! One minor comment about `as_encryption_key` but otherwise I think it's good to go.

File: src/paillier.rs (hunk 5 above)
Comment (marsella) [reply src/paillier.rs:76]: Yeah, that's fair, I just saw that lovely PR that demolishes all of this. I will put it back as it was. Made this change when I tried to add `half_n` as a field on the keys instead of a method.
